### ISSUE TITLE
Are table objects necessary?

### ISSUE THREAD

**Issue by theoreticalbts** [koinos/koinos-chain#19]
TLDR:  No.  Table objects are an unnecessary performance drain.

Every object exists in what I will call a "table" for the purposes of this ticket.  Table is similar to an SQL table.  Tables are used to achieve these objectives:

- (a) Avoid key collisions of objects within a single smart contract.
- (b) Implement access control where a smart contract is only allowed to access its own tables.
- (c) Avoid key collisions of objects from different smart contracts.

Actually (c) is redundant, as any implementation of (b) will necessarily also implement (c).

In EOS, tables are implemented in their own table object (`table_id_object`).  A user smart contract object is always indexed by a `(table_id, primary_key)` pair.  (Pretend for now that secondary keys aren't a concern, I will address them in a different ticket.)

EOS's particular implementation of tables has a couple large design issues:

- (1) (Performance) Every smart contract call must access at least one table object (except purely computational stateless methods)
- (2) (Encapsulation violation) The `table_id_object` must necessarily be implemented natively, however it contains `account_name` fields (specifically `code` and `payer`).

For (1), there is actually *no need for table objects to exist*.  This would be a great performance optimization as it will get rid of one object fetch per smart contract call.

How can we implement (a) (b) (c) without table objects?  All we need to do is specify table ID's so that all the information for collision avoidance and access control is self-contained within the table ID.  So effectively table ID can be a two-element struct of the form `(smart_contract_id, table_num)`, where `smart_contract_id` represents the smart contract and `table_num` represents a particular table (so a single smart contract can have multiple tables).

As far as (2) goes, we don't want to put any of the parts of the system "business model" into a native table.  Syscalls that implement "read table object" and "write table object" effectively get the `smart_contract_id` from a privileged VM register in the context.  We only need to protect this privileged VM register from being written in an unprivileged context ("unprivileged context" = "user specified smart contract," but the logic of when to change privilege levels should be implemented in a system smart contract).
